### PR TITLE
Add CIDR and User Agent processors

### LINKED ISSUE #4164: [DOC] Update Expression Syntax documentation for Data Prepper
**What do you want to do?**
 
- [x] Request a change to existing documentation
- [ ] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.
 Data Prepper added `function` support for expression syntax and added a few functions recently. We'd like to update the existing documentation (https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/) with a new `function` section.

In the data prepper repo, this is documented in the `Function` section of the expression syntax doc: https://github.com/opensearch-project/data-prepper/blob/main/docs/expression_syntax.md#function

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
Closes #4164 and #4163 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _data-prepper/pipelines/expression-syntax.md [hunk 1]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+Data Prepper supports the following built-in functions that can be used in an expression.
+
+### `length()`
+
+The `length()` function takes one argument of the JSON pointer type and returns the length of the value passed. For example, `length(/message)` returns a length of `10`, when a key message exists in the event and has a value of `1234567890`.
+
+### `hasTags()`
+
+The `hastags()` function takes one or more string type arguments and returns `true` if all the arguments passed are present in an event's tags. When an argument does not exist in the event's tags, the function returns `false`. For example, if you use the express `hasTags("tag1")` and the event contains `tag1`, Data Prepper returns `true`. If you use the same expression but the event also contains  `tag4` tag, Data Prepper returns `false`.
```
Comment (oeyh) [inline _data-prepper/pipelines/expression-syntax.md:233]: > If you use the same expression but the event **also** contains `tag4` tag, Data Prepper returns `false`.

Do you mean the event contains both `tag1` and `tag4`? 
If that's the case, `hasTags("tag1")` still returns `true`. `hasTags("tag4")` or `hasTags("tag1", "tag4")` also returns true. But `hasTags("tag2")` or `hasTags("tag1", "tag2")` would return false.

**Review by oeyh — COMMENTED** [review]
Looks good! Just one small issue:

File: _data-prepper/pipelines/expression-syntax.md [hunk 2]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+Data Prepper supports the following built-in functions that can be used in an expression.
+
+### `length()`
+
+The `length()` function takes one argument of the JSON pointer type and returns the length of the value passed. For example, `length(/message)` returns a length of `10`, when a key message exists in the event and has a value of `1234567890`.
+
+### `hasTags()`
+
+The `hastags()` function takes one or more string type arguments and returns `true` if all the arguments passed are present in an event's tags. When an argument does not exist in the event's tags, the function returns `false`. For example, if you use the expression `hasTags("tag1")` and the event contains `tag1`, Data Prepper returns `true`. If you use the expression `hasTags("tag2")` and the event but the event only contains a `tag1` tag, Data Prepper returns `false`.
```
Comment (oeyh) [inline _data-prepper/pipelines/expression-syntax.md:233]:
```suggestion
The `hastags()` function takes one or more string type arguments and returns `true` if all the arguments passed are present in an event's tags. When an argument does not exist in the event's tags, the function returns `false`. For example, if you use the expression `hasTags("tag1")` and the event contains `tag1`, Data Prepper returns `true`. If you use the expression `hasTags("tag2")` but the event only contains a `tag1` tag, Data Prepper returns `false`.
```

**Review by oeyh — APPROVED** [review]
Thanks for updating. Looks good on technical side.

File: _data-prepper/pipelines/configuration/processors/user-agent.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+If, for example, the `ua` field contains the following UA string:
+
+```json
+{
+  "ua":  "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1"
+}
+```
+
+The `user_agent` processors parses the string into a format compatible with Elastic Common Schema (ECS), then adds the result to the specified target, as showing in the following example:
```
Comment (cwillum) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:32]:
```suggestion
The `user_agent` processor parses the string into a format compatible with Elastic Common Schema (ECS), then adds the result to the specified target, as shown in the following example:
```

File: _data-prepper/pipelines/configuration/processors/user-agent.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+nav_order: 130
+---
+
+# user_agent
+
+The `user_agent` processor parses any user agent (UA) string in an event then adds the parsing results to the event's write data.
+
+## Usage
+
+In this example, the `user_agent` processor calls the source that contains the UA string, the `ua` field, and indicates the key where the parsed string will write, `user_agent`.
```
Comment (cwillum) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:15]: Is this first example of the `.yml` file illustrating the parsing of the user agent string in an event, as described in the first statement of the "# user_agent" definition above? And then does the example in JSON  illustrate the parsing of the results to the event's write data in the second statement in the "# user_agent" definition? If yes, can you somehow say that this first example performs a first step and the second example performs a second step in the # user_agent" definition? 

Or, is the .yml file example a configuration step while the JSON example fulfills both the parsing and the adding/writing process? 

I guess what I'm asking is better delineation between what the two examples are achieving.

File: _data-prepper/pipelines/expression-syntax.md (hunk 2 above)
Comment (cwillum) [inline _data-prepper/pipelines/expression-syntax.md:233]:
```suggestion
The `hastags()` function takes one or more string-type arguments and returns `true` if all the passed arguments are present in an event's tags. When an argument does not exist in the event's tags, the function returns `false`. For example, if you use the expression `hasTags("tag1")` and the event contains `tag1`, Data Prepper returns `true`. If you use the expression `hasTags("tag2")` and the event, and only the event, contains a `tag1` tag, Data Prepper returns `false`.
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 5]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+### `getMetadata()`
+
+The `getMetadata()` function takes one literal string argument to look up specific keys in a an event's metadata. If the key contains a `/`, then the function looks up the metadata recursively. When passed, the expression returns the value corresponding to the key. The value returned can be any type. For example, if the metadata contains `{"key1": "value2", "key2": 10}`, then the function, `getMetadata("key1")`, returns `value2`. The function, `getMetadata("key2")`, returns 10.
+
+### `contains()`
+
+The `contains()` function takes two string arguments and looks to see if either a literal string or a JSON pointer is contained within an event. When the second argument contains a substring of the first argument, such as `contains("abcde", "abcd")`, the function returns `true`. If the second argument does not contain any substrings, such as `contains("abcde", "xyz")`, returns `false`.
```
Comment (cwillum) [inline _data-prepper/pipelines/expression-syntax.md:241]:
```suggestion
The `contains()` function takes two string arguments and looks to see if either a literal string or a JSON pointer is contained within an event. When the second argument contains a substring of the first argument, such as `contains("abcde", "abcd")`, the function returns `true`. If the second argument does not contain any substrings, such as `contains("abcde", "xyz")`, it returns `false`.
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 6]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+### `cidrContains()`
+
+The `cidrContains()` function takes two or more arguments. The first argument is a JSON pointer which represents the key to the an IP address to check. It supports both IPv4 and IPv6 addresses. Every subsequent argument after the key is string type that represents CIDR blocks to check against.
```
Comment (cwillum) [inline _data-prepper/pipelines/expression-syntax.md:245]:
```suggestion
The `cidrContains()` function takes two or more arguments. The first argument is a JSON pointer, which represents the key to the IP address that is checked. It supports both IPv4 and IPv6 addresses. Every argument that comes after the key is a string type that represents CIDR blocks that are checked against.
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 7]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+If the IP address in the first argument is in the range of any of the given CIDR blocks, the function returns `true`. If the IP address is not in range of the CIDR blocks, the function returns false. For example, `cidrContains(/sourceIp,"192.0.2.0/24","10.0.1.0/16")` will returns as `true` if the `sourceIp` field indicated in the JSON pointed has a value of `192.0.2.5`.
```
Comment (cwillum) [inline _data-prepper/pipelines/expression-syntax.md:247]:
```suggestion
If the IP address in the first argument is in the range of any of the given CIDR blocks, the function returns `true`. If the IP address is not in the range of the CIDR blocks, the function returns false. For example, `cidrContains(/sourceIp,"192.0.2.0/24","10.0.1.0/16")` will return `true` if the `sourceIp` field indicated in the JSON pointer has a value of `192.0.2.5`.
```

**Review by cwillum — APPROVED** [review]
As always, suggestions only. Looks good.

File: _data-prepper/pipelines/configuration/processors/user-agent.md (hunk 4 above)
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/processors/user-agent.md]: This should be resolved now.

File: _data-prepper/pipelines/configuration/processors/user-agent.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+In this example, the `user_agent` processor calls the source that contains the UA string, the `ua` field, and indicates the key where the parsed string will write, `user_agent`, as shown in the following example:
+
+```yaml
+  processor:
+    - user_agent:
+        source: "ua"
+        target: "user_agent"
+```
+
+If, for example, the following event contain the `ua` field contains the following UA string:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:24]:
```suggestion
If, for example, the following event contains the `ua` field contains the following UA string:
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:24]: There are words missing in this sentence, either after "contains" or "field".

File: _data-prepper/pipelines/configuration/processors/user-agent.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+```json
+{
+  "ua":  "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.1 Mobile/15E148 Safari/604.1"
+}
+```
+
+The `user_agent` processors parses the string into a format compatible with Elastic Common Schema (ECS), then adds the result to the specified target, as shown in the following example:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:32]:
```suggestion
The `user_agent` processor parses the string into a format compatible with Elastic Common Schema (ECS) and then adds the result to the specified target, as shown in the following example:
```

File: _data-prepper/pipelines/configuration/processors/user-agent.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+## Configuration options
+
+You can use the following configuration options with the `user_agent` processor.
+
+| Option | Required | Description |
+| :--- | :--- | :--- |
+| `source` | Yes | The field in the event that will be parsed. 
+| `target` | No | The field where the parsed event will write. Default is `user_agent`.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:60]:
```suggestion
| `target` | No | The field to which the parsed event will write. Default is `user_agent`. 
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 11]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
 | `,`                  | Set value delimiter      | No                   | `/a in {200, 202}`<br>`/a in {200,202}`<br>`/a in {200 , 202}` | `/a in {200,}`                        |
+
+
+## Functions
+
+Data Prepper supports the following built-in functions that can be used in an expression.
+
+### `length()`
+
+The `length()` function takes one argument of the JSON pointer type and returns the length of the value passed. For example, `length(/message)` returns a length of `10`, when a key message exists in the event and has a value of `1234567890`.
```
Comment (natebower) [inline _data-prepper/pipelines/expression-syntax.md:229]:
```suggestion
The `length()` function takes one argument of the JSON pointer type and returns the length of the value passed. For example, `length(/message)` returns a length of `10` when a key message exists in the event and has a value of `1234567890`.
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 12]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+### `hasTags()`
+
+The `hastags()` function takes one or more string type arguments and returns `true` if all the arguments passed are present in an event's tags. When an argument does not exist in the event's tags, the function returns `false`. For example, if you use the expression `hasTags("tag1")` and the event contains `tag1`, Data Prepper returns `true`. If you use the expression `hasTags("tag2")` but the event only contains a `tag1` tag, Data Prepper returns `false`.
+
+### `getMetadata()`
+
+The `getMetadata()` function takes one literal string argument to look up specific keys in a an event's metadata. If the key contains a `/`, then the function looks up the metadata recursively. When passed, the expression returns the value corresponding to the key. The value returned can be any type. For example, if the metadata contains `{"key1": "value2", "key2": 10}`, then the function, `getMetadata("key1")`, returns `value2`. The function, `getMetadata("key2")`, returns 10.
```
Comment (natebower) [inline _data-prepper/pipelines/expression-syntax.md:237]:
```suggestion
The `getMetadata()` function takes one literal string argument to look up specific keys in a an event's metadata. If the key contains a `/`, then the function looks up the metadata recursively. When passed, the expression returns the value corresponding to the key. The value returned can be of any type. For example, if the metadata contains `{"key1": "value2", "key2": 10}`, then the function, `getMetadata("key1")`, returns `value2`. The function, `getMetadata("key2")`, returns 10.
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 13]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+### `contains()`
+
+The `contains()` function takes two string arguments and looks to see if either a literal string or a JSON pointer is contained within an event. When the second argument contains a substring of the first argument, such as `contains("abcde", "abcd")`, the function returns `true`. If the second argument does not contain any substrings, such as `contains("abcde", "xyz")`, it returns `false`.
```
Comment (natebower) [inline _data-prepper/pipelines/expression-syntax.md:241]:
```suggestion
The `contains()` function takes two string arguments and determines whether either a literal string or a JSON pointer is contained within an event. When the second argument contains a substring of the first argument, such as `contains("abcde", "abcd")`, the function returns `true`. If the second argument does not contain any substrings, such as `contains("abcde", "xyz")`, it returns `false`.
```

File: _data-prepper/pipelines/expression-syntax.md [hunk 14]
```diff
@@ -208,3 +218,30 @@ White space is **required** surrounding set initializers, priority expressions,
...
+
+### `cidrContains()`
+
+The `cidrContains()` function takes two or more arguments. The first argument is a JSON pointer, which represents the key to the IP address that is checked. It supports both IPv4 and IPv6 addresses. Every argument that comes after the key is a string type that represents CIDR blocks that are checked against.
+
+If the IP address in the first argument is in the range of any of the given CIDR blocks, the function returns `true`. If the IP address is not in the range of the CIDR blocks, the function returns false. For example, `cidrContains(/sourceIp,"192.0.2.0/24","10.0.1.0/16")` will return `true` if the `sourceIp` field indicated in the JSON pointer has a value of `192.0.2.5`.
```
Comment (natebower) [inline _data-prepper/pipelines/expression-syntax.md:247]:
```suggestion
If the IP address in the first argument is in the range of any of the given CIDR blocks, the function returns `true`. If the IP address is not in the range of the CIDR blocks, the function returns `false`. For example, `cidrContains(/sourceIp,"192.0.2.0/24","10.0.1.0/16")` will return `true` if the `sourceIp` field indicated in the JSON pointer has a value of `192.0.2.5`.
```

**Review by natebower — CHANGES_REQUESTED** [review]
@Naarcha-AWS Tag me on the sentence with a comment once you've revised it so I can approve. Otherwise, all minor changes. Thanks!

File: _data-prepper/pipelines/configuration/processors/user-agent.md (hunk 8 above)
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/user-agent.md:24]: @natebower: How does this sound?
```suggestion
The following example event contains the `ua` field with the a string that details information about a user: 
```

Comment (natebower) [reply _data-prepper/pipelines/configuration/processors/user-agent.md]: How about that?

**Review by natebower — APPROVED** [review]
LGTM :-)

**Pull request merged**